File: backend/swift/rt.swift
// AgentS runtime for the Swift backend.
//
// Strings are measured and sliced in Unicode scalars, not in Characters: Swift
// counts grapheme clusters where Rust's `chars()` and Python's `len()` count
// scalars, and a backend that disagreed there would fail the differential gate
// on any combining sequence.
//
// Foundation is deliberately not imported. Trim, replace and search are the only
// operations that would need it, they are a dozen lines each, and the generated
// code stays free of any dependency the target platform has to supply.

/// `Result` is hand-rolled because the stdlib's constrains its failure type to
/// `Error`, and AgentS writes `(Result Int64 String)`. Conforming `String` to
/// `Error` retroactively would change an imported type for every module that
/// links this runtime.
public enum ASResult<T, E> {
    case ok(T)
    case err(E)
}

extension ASResult: Equatable where T: Equatable, E: Equatable {}

/// `Pair` is a struct rather than a tuple: Swift tuples cannot conform to
/// protocols, which would leave `list-contains?`, `list-sort` and
/// `map-from-pairs` unusable over pairs.
public struct ASPair<A, B> {
    public let first: A
    public let second: B
    public init(_ a: A, _ b: B) { first = a; second = b }
}

extension ASPair: Equatable where A: Equatable, B: Equatable {}
extension ASPair: Hashable where A: Hashable, B: Hashable {}
extension ASPair: Comparable where A: Comparable, B: Comparable {
    public static func < (l: Self, r: Self) -> Bool {
        if l.first != r.first { return l.first < r.first }
        return l.second < r.second
    }
}

/// Carries an AgentS `err` value out of a `try`. Typed throws let the enclosing
/// function catch exactly this and nothing else, so `try` propagates without the
/// caller-visible colouring that a `throws` signature would impose.
public struct ASThrown<E>: Error {
    public let value: E
    public init(_ value: E) { self.value = value }
}

public enum RT {
    // MARK: arithmetic

    public static func div<T: BinaryInteger>(_ a: T, _ b: T) -> T {
        precondition(b != 0, "division by zero")
        return a / b
    }
    public static func div(_ a: Double, _ b: Double) -> Double { a / b }

    public static func rem<T: BinaryInteger>(_ a: T, _ b: T) -> T {
        precondition(b != 0, "modulo by zero")
        return a % b
    }
    public static func rem(_ a: Double, _ b: Double) -> Double { a.truncatingRemainder(dividingBy: b) }

    public static func checkedDiv<T: BinaryInteger>(_ a: T, _ b: T) -> T? { b == 0 ? nil : a / b }
    public static func checkedRem<T: BinaryInteger>(_ a: T, _ b: T) -> T? { b == 0 ? nil : a % b }

    // MARK: strings

    public static func strLen(_ s: String) -> Int64 { Int64(s.unicodeScalars.count) }
    public static func concat(_ parts: [String]) -> String { parts.joined() }
    public static func chars(_ s: String) -> [String] { s.unicodeScalars.map { String($0) } }
    public static func strRev(_ s: String) -> String { fromScalars(Array(s.unicodeScalars).reversed()) }

    public static func trim(_ s: String) -> String {
        var u = Array(s.unicodeScalars)
        while let f = u.first, isSpace(f) { u.removeFirst() }
        while let l = u.last, isSpace(l) { u.removeLast() }
        return fromScalars(u)
    }

    public static func contains(_ s: String, _ sub: String) -> Bool { scalarIndex(s, sub) != nil }

    /// An empty separator yields a leading and trailing empty field, as Rust's
    /// `str::split("")` does. Python raises there instead; the divergence is the
    /// specification's, not this backend's.
    public static func split(_ s: String, _ sep: String) -> [String] {
        let u = Array(s.unicodeScalars)
        let p = Array(sep.unicodeScalars)
        if p.isEmpty { return [""] + u.map { String($0) } + [""] }
        var out: [String] = []
        var start = 0
        var i = 0
        while i + p.count <= u.count {
            if Array(u[i..<(i + p.count)]) == p {
                out.append(fromScalars(Array(u[start..<i])))
                i += p.count
                start = i
            } else {
                i += 1
            }
        }
        out.append(fromScalars(Array(u[start...])))
        return out
    }

    public static func replace(_ s: String, _ from: String, _ to: String) -> String {
        from.isEmpty ? s : split(s, from).joined(separator: to)
    }

    public static func strSlice(_ s: String, _ a: Int64, _ b: Int64) -> String? {
        let u = Array(s.unicodeScalars)
        guard a >= 0, b >= a, b <= Int64(u.count) else { return nil }
        return fromScalars(Array(u[Int(a)..<Int(b)]))
    }

    public static func strIndexOf(_ s: String, _ sub: String) -> Int64? {
        scalarIndex(s, sub).map(Int64.init)
    }

    /// Matches Rust's `{:?}` and Python's `repr` on a float: shortest form that
    /// round-trips, with a decimal point always present.
    public static func fmtF64(_ x: Double) -> String {
        let s = String(x)
        return s.contains(".") || s.contains("e") || s.contains("n") || s.contains("i") ? s : s + ".0"
    }

    public static func toI64(_ s: String) -> Int64? { Int64(trim(s)) }
    public static func toF64(_ s: String) -> Double? { Double(trim(s)) }
    public static func toI32(_ n: Int64) -> Int32? { Int32(exactly: n) }

    /// Out-of-range saturates rather than failing, which is what the Rust
    /// backend's `as i64` cast does. Only NaN and infinity are `none`.
    public static func fToI(_ x: Double) -> Int64? {
        if x.isNaN || x.isInfinite { return nil }
        let t = x.rounded(.towardZero)
        if t <= -9223372036854775808.0 { return Int64.min }
        if t >= 9223372036854775807.0 { return Int64.max }
        return Int64(t)
    }

    private static func isSpace(_ u: Unicode.Scalar) -> Bool {
        u == " " || u == "\t" || u == "\n" || u == "\r" || u == "\u{0B}" || u == "\u{0C}"
    }

    private static func fromScalars(_ u: [Unicode.Scalar]) -> String {
        String(String.UnicodeScalarView(u))
    }

    private static func scalarIndex(_ s: String, _ sub: String) -> Int? {
        let u = Array(s.unicodeScalars)
        let p = Array(sub.unicodeScalars)
        if p.isEmpty { return 0 }
        if p.count > u.count { return nil }
        for i in 0...(u.count - p.count) where Array(u[i..<(i + p.count)]) == p { return i }
        return nil
    }

    // MARK: lists

    public static func at<T>(_ xs: [T], _ i: Int64) -> T? {
        i >= 0 && i < Int64(xs.count) ? xs[Int(i)] : nil
    }
    public static func tail<T>(_ xs: [T]) -> [T]? { xs.isEmpty ? nil : Array(xs.dropFirst()) }

    public static func listSlice<T>(_ xs: [T], _ a: Int64, _ b: Int64) -> [T]? {
        guard a >= 0, b >= a, b <= Int64(xs.count) else { return nil }
        return Array(xs[Int(a)..<Int(b)])
    }

    public static func indexOf<T: Equatable>(_ xs: [T], _ x: T) -> Int64? {
        xs.firstIndex(of: x).map(Int64.init)
    }

    public static func sortBy<T, K: Comparable>(_ f: (T) -> K, _ xs: [T]) -> [T] {
        xs.sorted { f($0) < f($1) }
    }

    public static func range(_ a: Int64, _ b: Int64) -> [Int64] { a >= b ? [] : Array(a..<b) }

    public static func zip<A, B>(_ a: [A], _ b: [B]) -> [ASPair<A, B>] {
        Swift.zip(a, b).map { ASPair($0, $1) }
    }

    public static func sum<T: AdditiveArithmetic>(_ xs: [T]) -> T { xs.reduce(.zero, +) }

    // MARK: maps
    //
    // Iteration is sorted by key. The specification orders map-keys/values/pairs,
    // Swift's Dictionary is unordered, and an unspecified order would make the
    // backends disagree on identical input.

    public static func mGet<K: Hashable, V>(_ m: [K: V], _ k: K) -> V? { m[k] }

    public static func mSet<K: Hashable, V>(_ m: [K: V], _ k: K, _ v: V) -> [K: V] {
        var out = m
        out[k] = v
        return out
    }

    public static func mDel<K: Hashable, V>(_ m: [K: V], _ k: K) -> [K: V] {
        var out = m
        out.removeValue(forKey: k)
        return out
    }

    public static func mKeys<K: Hashable & Comparable, V>(_ m: [K: V]) -> [K] { m.keys.sorted() }

    public static func mValues<K: Hashable & Comparable, V>(_ m: [K: V]) -> [V] {
        m.keys.sorted().map { m[$0]! }
    }

    public static func mPairs<K: Hashable & Comparable, V>(_ m: [K: V]) -> [ASPair<K, V>] {
        m.keys.sorted().map { ASPair($0, m[$0]!) }
    }

    public static func mFrom<K: Hashable, V>(_ ps: [ASPair<K, V>]) -> [K: V] {
        var out: [K: V] = [:]
        for p in ps { out[p.first] = p.second }
        return out
    }

    // MARK: option and result

    public static func isOk<T, E>(_ r: ASResult<T, E>) -> Bool {
        if case .ok = r { return true }
        return false
    }

    public static func resOr<T, E>(_ r: ASResult<T, E>, _ d: T) -> T {
        if case .ok(let v) = r { return v }
        return d
    }

    public static func resMap<A, B, E>(_ f: (A) -> B, _ r: ASResult<A, E>) -> ASResult<B, E> {
        switch r {
        case .ok(let v): return .ok(f(v))
        case .err(let e): return .err(e)
        }
    }

    public static func resMapErr<T, E, F>(_ f: (E) -> F, _ r: ASResult<T, E>) -> ASResult<T, F> {
        switch r {
        case .ok(let v): return .ok(v)
        case .err(let e): return .err(f(e))
        }
    }

    public static func optToRes<T, E>(_ o: T?, _ e: E) -> ASResult<T, E> {
        if let v = o { return .ok(v) }
        return .err(e)
    }

    public static func resToOpt<T, E>(_ r: ASResult<T, E>) -> T? {
        if case .ok(let v) = r { return v }
        return nil
    }

    /// The `try` form: yields the ok value, or throws the err value to the
    /// enclosing function's catch.
    public static func unwrap<T, E>(_ r: ASResult<T, E>) throws(ASThrown<E>) -> T {
        switch r {
        case .ok(let v): return v
        case .err(let e): throw ASThrown(e)
        }
    }
}

File: backend/rust/rt.rs
//! AgentScript runtime for the Rust backend.
//!
//! Map lowers to BTreeMap, not HashMap: the language specifies sorted iteration
//! for map-keys/values/pairs, and an unspecified order would make the backends
//! disagree on identical input — which the differential harness would then report
//! as a transpiler defect rather than as the specification gap it is.
#![allow(dead_code, unused_imports)]
use std::collections::BTreeMap;

// `N` spans Int32, Int64 and Float64, so a numeric helper fixed to one of them is
// a backend narrower than the language it claims to implement.
//
// Every integer operation goes through a `checked_*` even though debug rustc
// would trap on the bare operator: trapping is what the language specifies, and
// tying it to a compilation profile would make `-O` a different language.
pub trait Num: Copy + PartialEq {
    const ZERO: Self;
    /// `None` when the quotient is not representable, which for two's complement
    /// is `MIN / -1` alone.
    fn quot(self, b: Self) -> Option<Self>;
    fn rest(self, b: Self) -> Self;
    fn plus(self, b: Self) -> Self;
    fn minus(self, b: Self) -> Self;
    fn times(self, b: Self) -> Self;
    fn negate(self) -> Self;
    fn magnitude(self) -> Self;
}

macro_rules! int_num {
    ($t:ty) => {
        impl Num for $t {
            const ZERO: $t = 0;
            fn quot(self, b: $t) -> Option<$t> { self.checked_div(b) }
            // `MIN % -1` is 0, which is representable, so the remainder wraps
            // where the quotient traps.
            fn rest(self, b: $t) -> $t { self.wrapping_rem(b) }
            fn plus(self, b: $t) -> $t { self.checked_add(b).expect("overflow in addition") }
            fn minus(self, b: $t) -> $t { self.checked_sub(b).expect("overflow in subtraction") }
            fn times(self, b: $t) -> $t { self.checked_mul(b).expect("overflow in multiplication") }
            fn negate(self) -> $t { self.checked_neg().expect("overflow in negation") }
            fn magnitude(self) -> $t { self.checked_abs().expect("overflow in absolute value") }
        }
    };
}

int_num!(i32);
int_num!(i64);

impl Num for f64 {
    const ZERO: f64 = 0.0;
    fn quot(self, b: f64) -> Option<f64> { Some(self / b) }
    fn rest(self, b: f64) -> f64 { self % b }
    fn plus(self, b: f64) -> f64 { self + b }
    fn minus(self, b: f64) -> f64 { self - b }
    fn times(self, b: f64) -> f64 { self * b }
    fn negate(self) -> f64 { -self }
    fn magnitude(self) -> f64 { self.abs() }
}

pub fn add<T: Num>(a: T, b: T) -> T { a.plus(b) }
pub fn sub<T: Num>(a: T, b: T) -> T { a.minus(b) }
pub fn mul<T: Num>(a: T, b: T) -> T { a.times(b) }
pub fn neg<T: Num>(a: T) -> T { a.negate() }
pub fn abs<T: Num>(a: T) -> T { a.magnitude() }

pub fn div<T: Num>(a: T, b: T) -> T {
    if b == T::ZERO { panic!("division by zero") }
    a.quot(b).expect("overflow in division")
}
pub fn rem<T: Num>(a: T, b: T) -> T {
    if b == T::ZERO { panic!("modulo by zero") }
    a.rest(b)
}
pub fn checked_div<T: Num>(a: T, b: T) -> Option<T> {
    if b == T::ZERO { None } else { a.quot(b) }
}
pub fn checked_rem<T: Num>(a: T, b: T) -> Option<T> {
    if b == T::ZERO { None } else { Some(a.rest(b)) }
}

/// Holds a NaN somewhere. A value that does is not comparable with *itself*,
/// which finds it without knowing T's shape — and T's shape is not available,
/// because `list-sort` is declared over every type.
fn unordered<T: PartialOrd>(x: &T) -> bool { x.partial_cmp(x).is_none() }

/// The language's sort order: a value holding a NaN sorts after every value that
/// does not, and NaN-holding values tie with one another so a stable sort leaves
/// them in input order.
///
/// The NaN test comes first rather than as a fallback for an incomparable pair,
/// which is what makes this a transitive order: `[0.0, nan]` compares Less than
/// `[0.5]` on its first element while `[nan]` compares Greater, so a rule that
/// asked `partial_cmp` first would sort two values that tie with each other onto
/// opposite sides of a third — and Rust's sort detects that and panics.
fn nan_last<T: PartialOrd>(a: &T, b: &T) -> std::cmp::Ordering {
    use std::cmp::Ordering;
    match (unordered(a), unordered(b)) {
        (true, true) => Ordering::Equal,
        (true, false) => Ordering::Greater,
        (false, true) => Ordering::Less,
        // Neither holds a NaN, so the partial order is total over the two.
        (false, false) => a.partial_cmp(b).expect("incomparable values in a sort"),
    }
}
fn sorts_before<T: PartialOrd>(a: &T, b: &T) -> bool {
    nan_last(a, b) == std::cmp::Ordering::Less
}
// Selection follows the same order as the sort, so `min` is the head of
// `list-sort` and not a separate rule: keep the first unless the second sorts
// strictly before it.
pub fn min<T: PartialOrd>(a: T, b: T) -> T { if sorts_before(&b, &a) { b } else { a } }
pub fn max<T: PartialOrd>(a: T, b: T) -> T { if sorts_before(&a, &b) { b } else { a } }

pub fn str_len(s: &str) -> i64 { s.chars().count() as i64 }
pub fn concat(parts: &[String]) -> String { parts.concat() }
pub fn chars(s: &str) -> Vec<String> { s.chars().map(|c| c.to_string()).collect() }
pub fn str_rev(s: &str) -> String { s.chars().rev().collect() }
pub fn split(s: &str, sep: &str) -> Vec<String> {
    s.split(sep).map(|x| x.to_string()).collect()
}
pub fn str_slice(s: &str, a: i64, b: i64) -> Option<String> {
    let n = s.chars().count() as i64;
    if a < 0 || b < a || b > n { return None }
    Some(s.chars().skip(a as usize).take((b - a) as usize).collect())
}
pub fn str_index_of(s: &str, sub: &str) -> Option<i64> {
    s.find(sub).map(|byte| s[..byte].chars().count() as i64)
}
pub fn fmt_f64(x: f64) -> String {
    // string-from-float64 has one meaning, and Rust's Debug spells three things
    // differently from Python's repr: NaN's case, an unsigned exponent, and an
    // exponent under two digits. Everything else already round-trips shortest.
    if x.is_nan() { return "nan".to_string() }
    let s = format!("{:?}", x);
    match s.split_once('e') {
        Some((mantissa, exp)) => {
            let (sign, digits) = match exp.strip_prefix('-') {
                Some(d) => ("-", d),
                None => ("+", exp),
            };
            format!("{}e{}{:0>2}", mantissa, sign, digits)
        }
        None => s,
    }
}
// The width is in the signature, not left to inference: an integer literal with
// nothing else constraining it falls back to i32 in Rust, so `{0}.to_string()`
// and `({0} as f64)` stopped compiling the moment the argument was a literal
// outside i32 — including Int64::MIN, which is a value the language has.
pub fn fmt_i64(n: i64) -> String { n.to_string() }
pub fn i_to_f(n: i64) -> f64 { n as f64 }
pub fn to_i64(s: &str) -> Option<i64> { s.trim().parse().ok() }
pub fn to_f64(s: &str) -> Option<f64> { s.trim().parse().ok() }
pub fn to_i32(n: i64) -> Option<i32> { i32::try_from(n).ok() }
pub fn f_to_i(x: f64) -> Option<i64> {
    // `as` saturates rather than failing, so the range decides before the cast:
    // 1e30 would otherwise answer i64::MAX and call that a conversion. NaN and
    // both infinities fall out of the same two comparisons.
    let t = x.trunc();
    if t >= -9223372036854775808.0 && t < 9223372036854775808.0 {
        Some(t as i64)
    } else {
        None
    }
}

pub fn at<T: Clone>(xs: &[T], i: i64) -> Option<T> {
    if i < 0 { return None }
    xs.get(i as usize).cloned()
}
pub fn tail<T: Clone>(xs: &[T]) -> Option<Vec<T>> {
    if xs.is_empty() { None } else { Some(xs[1..].to_vec()) }
}
pub fn cons<T>(x: T, xs: Vec<T>) -> Vec<T> {
    let mut v = vec![x];
    v.extend(xs);
    v
}
pub fn append<T>(mut a: Vec<T>, b: Vec<T>) -> Vec<T> { a.extend(b); a }
pub fn rev<T>(mut xs: Vec<T>) -> Vec<T> { xs.reverse(); xs }
pub fn list_slice<T: Clone>(xs: &[T], a: i64, b: i64) -> Option<Vec<T>> {
    let n = xs.len() as i64;
    if a < 0 || b < a || b > n { return None }
    Some(xs[a as usize..b as usize].to_vec())
}
pub fn index_of<T: PartialEq>(xs: &[T], x: &T) -> Option<i64> {
    xs.iter().position(|y| y == x).map(|i| i as i64)
}
pub fn sort<T: PartialOrd>(mut xs: Vec<T>) -> Vec<T> {
    xs.sort_by(nan_last);
    xs
}
// The language's higher-order arguments take their element by value, so the
// helpers do too: a closure written against `&T` here could never be the same
// closure the source declares.
pub fn sort_by<T: Clone, K: PartialOrd, F: Fn(T) -> K>(mut xs: Vec<T>, f: F) -> Vec<T> {
    xs.sort_by(|a, b| nan_last(&f(a.clone()), &f(b.clone())));
    xs
}
// Passing the predicate to a generic function, rather than invoking a closure
// literal inline, is what lets rustc infer an elided parameter type: the bound
// supplies the expected signature the same way the checker's does.
pub fn filter<T: Clone, F: Fn(T) -> bool>(xs: Vec<T>, f: F) -> Vec<T> {
    xs.into_iter().filter(|x| f(x.clone())).collect()
}
pub fn range(a: i64, b: i64) -> Vec<i64> { if a >= b { vec![] } else { (a..b).collect() } }
pub fn zip<A, B>(a: Vec<A>, b: Vec<B>) -> Vec<(A, B)> { a.into_iter().zip(b).collect() }
pub fn sum<T: Num>(xs: Vec<T>) -> T { xs.into_iter().fold(T::ZERO, Num::plus) }
pub fn least<T: PartialOrd + Clone>(xs: &[T]) -> Option<T> {
    xs.iter().cloned().reduce(|a, b| if sorts_before(&b, &a) { b } else { a })
}
pub fn greatest<T: PartialOrd + Clone>(xs: &[T]) -> Option<T> {
    xs.iter().cloned().reduce(|a, b| if sorts_before(&a, &b) { b } else { a })
}

pub fn m_get<K: Ord, V: Clone>(m: &BTreeMap<K, V>, k: &K) -> Option<V> { m.get(k).cloned() }
pub fn m_set<K: Ord, V>(mut m: BTreeMap<K, V>, k: K, v: V) -> BTreeMap<K, V> {
    m.insert(k, v);
    m
}
pub fn m_del<K: Ord, V>(mut m: BTreeMap<K, V>, k: &K) -> BTreeMap<K, V> { m.remove(k); m }
pub fn m_pairs<K: Ord + Clone, V: Clone>(m: &BTreeMap<K, V>) -> Vec<(K, V)> {
    m.iter().map(|(k, v)| (k.clone(), v.clone())).collect()
}
pub fn m_from<K: Ord, V>(ps: Vec<(K, V)>) -> BTreeMap<K, V> { ps.into_iter().collect() }

// ---------- I/O ----------
// The case is chosen from errno where one exists, and from ErrorKind otherwise,
// because the Python runtime reaches its case from errno and the two have to
// agree for the same condition. The differential gate compares them.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum IoError {
    NotFound,
    PermissionDenied,
    AlreadyExists,
    InvalidPath,
    Interrupted,
    Other,
}

impl IoError {
    pub fn case(&self) -> &'static str {
        match self {
            IoError::NotFound => "not-found",
            IoError::PermissionDenied => "permission-denied",
            IoError::AlreadyExists => "already-exists",
            IoError::InvalidPath => "invalid-path",
            IoError::Interrupted => "interrupted",
            IoError::Other => "other",
        }
    }
}

fn io_err(e: std::io::Error) -> IoError {
    match e.raw_os_error() {
        Some(2) => IoError::NotFound,
        Some(13) => IoError::PermissionDenied,
        Some(17) => IoError::AlreadyExists,
        Some(20) | Some(21) => IoError::InvalidPath,
        Some(4) => IoError::Interrupted,
        _ => match e.kind() {
            std::io::ErrorKind::NotFound => IoError::NotFound,
            std::io::ErrorKind::PermissionDenied => IoError::PermissionDenied,
            std::io::ErrorKind::AlreadyExists => IoError::AlreadyExists,
            std::io::ErrorKind::Interrupted => IoError::Interrupted,
            _ => IoError::Other,
        },
    }
}

pub fn read_line() -> Result<Option<String>, IoError> {
    let mut s = String::new();
    match std::io::BufRead::read_line(&mut std::io::stdin().lock(), &mut s) {
        Ok(0) => Ok(None),
        Ok(_) => Ok(Some(s.trim_end_matches('\n').to_string())),
        Err(e) => Err(io_err(e)),
    }
}

pub fn read_all() -> Result<String, IoError> {
    let mut s = String::new();
    std::io::Read::read_to_string(&mut std::io::stdin(), &mut s)
        .map(|_| s)
        .map_err(io_err)
}

fn write_to<W: std::io::Write>(mut w: W, text: &str) -> Result<(), IoError> {
    w.write_all(text.as_bytes()).map_err(io_err)?;
    w.flush().map_err(io_err)
}

pub fn print_out(s: &str) -> Result<(), IoError> { write_to(std::io::stdout(), s) }
pub fn println(s: &str) -> Result<(), IoError> { write_to(std::io::stdout(), &format!("{}\n", s)) }
pub fn eprintln(s: &str) -> Result<(), IoError> { write_to(std::io::stderr(), &format!("{}\n", s)) }

pub fn file_read(path: &str) -> Result<String, IoError> {
    std::fs::read_to_string(path).map_err(io_err)
}

pub fn file_write(path: &str, text: &str) -> Result<(), IoError> {
    std::fs::write(path, text).map_err(io_err)
}

pub fn file_append(path: &str, text: &str) -> Result<(), IoError> {
    let mut f = std::fs::OpenOptions::new()
        .create(true).append(true).open(path).map_err(io_err)?;
    write_to(&mut f, text)
}

pub fn file_exists(path: &str) -> Result<bool, IoError> {
    Ok(std::path::Path::new(path).exists())
}

/// Host entry glue: a program's Result becomes its exit status. `err` prints the
/// case name, which is the only part of a failure the language defines.
pub fn main_exit(r: Result<(), IoError>) -> i32 {
    match r {
        Ok(()) => 0,
        Err(e) => { let _ = eprintln(e.case()); 1 }
    }
}
